extensions/amp-story/1.0/amp-story-quiz.js

```diff
@@ -0,0 +1,182 @@
+      throw new Error(e);
+    });
+    this.attachOptionActionHandlers_();
+
+    // TODO: ATTACH SURFACE OPTIONS USING CSS CUSTOM PROPERTIES
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @private
+   * Finds the prompt and options content
+   * and adds it to the shadow DOM.
+   *
+   * @param {Function} handleError
+   */
+  attachContent_(handleError) {
+    // TODO: If prompt is optional I need to handle the case where it's not present
+    const prompt = this.element.children[0];
+    // First child must be heading h1-h3
+    if (
+      !(prompt instanceof HTMLHeadingElement) ||
+      prompt.tagName[prompt.tagName.length - 1] > 3
+    ) {
+      handleError(
+        'The first child must be a heading element <h1>, <h2>, or <h3>'
+      );
+    }
+
+    prompt.setAttribute('class', 'i-amp-story-quiz-prompt');
+
+    const options = Array.from(this.element.querySelectorAll('option'));
+    if (options.length < 2 || options.length > 4) {
+      handleError('Improper number of options');
+    }
+
+    this.shadowRoot_
+      .querySelector('.i-amp-story-quiz-prompt-container')
+      .appendChild(prompt);
+
+    options.forEach((option, index) => this.configureOption_(option, index));
+
+    if (this.element.children.length !== 0) {
+      handleError('Too many children');
+    }
+  }
+
+  /**
+   * @private
+   * Creates an option container with option content,
+   * adds styling and answer choices,
+   * and adds it to the shadow DOM.
+   * @param {HTMLOptionElement} option
+   * @param {number} index
+   */
+  configureOption_(option, index) {
+    // Transfer the option information into a span -
+    // this allows the option container to house other markup,
+    // such as the answer choices
+    const convertedOption = document.createElement('span');
+    convertedOption.textContent = option.textContent;
+    if (option.hasAttribute('correct')) {
+      convertedOption.setAttribute('correct', 'correct');
+    }
+    option.remove();
+
+    // Set up the class and add it to the shadow root
+    convertedOption.setAttribute('class', 'i-amp-story-quiz-option');
+    this.shadowRoot_
+      .querySelector('.i-amp-story-quiz-option-container')
+      .appendChild(convertedOption);
+
+    // Create a container for the answer choice and add it to the shadow root
+    const answerChoice = document.createElement('span');
+    answerChoice.textContent = answerChoiceOptions[index];
+    answerChoice.setAttribute('class', 'i-amp-story-quiz-answer-choice');
+    convertedOption.prepend(answerChoice);
+  }
+
+  /**
+   * @private
+   * Attaches functions to each option to handle state transition.
+   *
+   */
+  attachOptionActionHandlers_() {
+    const options = Array.from(
+      this.shadowRoot_.querySelectorAll('.i-amp-story-quiz-option')
+    );
+
+    // attach listeners
+    options.forEach(option => {
+      option.addEventListener('click', () => {
```

Comment:
Two ideas so we don't have to set one click listener per option. Not sure which one is the best:

1. Use event delegation. By setting a click listener higher in the DOM and relying on event bubbling (or you could actually use capture), you should be able to retrieve the click target with only one listener.
2. Use AMP actions: https://amp.dev/documentation/guides-and-tutorials/learn/amp-actions-and-events/

What I don't really like with (1) is that we'd have to start/stop listening to events. Using (2) sounds better but I don't know how well it will work with stories out of the box. The great point is that we'd rely on the AMP events system and we would not need any additional listeners.

Please lmk if you have additional thoughts or questions :))

Reply:
I think option 2 could work, we could set it up such that tapping on the option container adds the `i-amp-story-quiz-option-post-selection` to each option, and tapping on an option adds the `i-amp-story-quiz-option-selected` class, then use CSS directly to replace the answer choices with icons.

I'll look into this.

Reply:
Done